Envoy core changes for reverse connections

basundhara-c opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: This commit collates the envoy core changes for reverse connections, described in this [github issue](https://github.com/envoyproxy/envoy/issues/33320). A detailed description of reverse connections concepts and workflows is provided in both the github issue and in the [examples section](https://github.com/envoyproxy/envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/examples/reverse_connection/README.md).

Additional Description: This PR involves several working components, that are added as part of the following extensions:
- "envoy.filters.http.reverse_conn" HTTP filter: https://github.com/envoyproxy/envoy/pull/37822
- "envoy.bootstrap.reverse_connection" bootstrap extension: https://github.com/envoyproxy/envoy/pull/37819
- "envoy.filters.listener.reverse_connection" listener filter: https://github.com/envoyproxy/envoy/pull/37821
- "envoy.reverse_connection.reverse_connection_listener_config" listener that initiates reverse connections: https://github.com/envoyproxy/envoy/pull/37820
- "envoy.reverse_connection.conn_pool": A custom conn pool that enables requests to be sent over reverse connections: https://github.com/envoyproxy/envoy/pull/37931
- "envoy.clusters.reverse_connection": A custom cluster type for downstream services that are behind reverse connections: https://github.com/envoyproxy/envoy/pull/37932

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

Signed-off-by: Basundhara Chakrabarty <basundhara.c@nutanix.com>
Co-authored-by: Arun Vasudevan <arun.vasudevan@nutanix.com>
Co-authored-by: Tejas Sangol <tejas.sangol@nutanix.com>
Co-authored-by: Aditya Jaltade <aditya.jaltade@nutanix.com>

alyssawilk commented on envoy/event/dispatcher.h:
```diff
@@ -286,6 +289,28 @@ class Dispatcher : public DispatcherBase, public ScopeTracker {
+   * Provides filters access to connection handler to save outgoing connections as
+   * incoming connections for reverse tunnels
+   */
+  virtual void setConnectionHandler(Network::ConnectionHandler* connection_handler) PURE;
```
I don't see what this function has to do with the dispatcher - did you stash this here as a convenient way to get the handler local to your thread?  I think you want to look into the thread local state (tls) getters and setters.

alyssawilk commented on envoy/network/connection_handler.h:
```diff
@@ -19,6 +21,26 @@
 namespace Network {
 
+// The thread local registry.
+class LocalRevConnRegistry {
```
I suspect you should be able to do this PR with minimal APIs in core Envoy, but instead in your extension directory

basundhara-c replied:
@alyssawilk  The main set of changes is to support the creation of the ReverseConnectionManager and ReverseConnectionHandler [thread-locally](https://github.com/nutanix/envoyproxy-envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/source/extensions/bootstrap/reverse_connection/reverse_conn_thread_local_registry.cc) as implemented using a thread-local registry. The ReverseConnectionManager and ReverseConnectionHandler need to come up [right after the workers are created](https://github.com/nutanix/envoyproxy-envoy/blob/457e1e985a0ef2c0812f809d34aa3add2d029433/source/common/listener_manager/listener_manager_impl.cc#L371) because they are involved in accepting reverse connections from other initiating envoy instances even if the local envoy is not initiating any reverse connections, as long as the [local envoy has the reverse connections bootstrap extension enabled](https://github.com/nutanix/envoyproxy-envoy/blob/457e1e985a0ef2c0812f809d34aa3add2d029433/source/server/server.cc#L820).

So, currently, we check if the reverse connection bootstrap extension is enabled by checking if the singleton is present, and if so, [right after workers are created, we post to each worker's dispatcher](https://github.com/nutanix/envoyproxy-envoy/blob/457e1e985a0ef2c0812f809d34aa3add2d029433/source/server/server.cc#L820) a functor to create the Thread Local Registries (creating the ReverseConnectionManager and ReverseConnectionHandler). This thread local registry is then parked with the Connection Handler and is accessed later wherever we need access to the  ReverseConnectionManager and ReverseConnectionHandler like [here,to initiate reverse conns](https://github.com/nutanix/envoyproxy-envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/source/extensions/reverse_connection/active_reverse_connection_listener.cc#L44), [here in the reverse conn filter, which accepts reverse conns](https://github.com/nutanix/envoyproxy-envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/source/extensions/filters/http/reverse_conn/reverse_conn_filter.h#L124) and in multiple other places where only the thread local dispatcher is otherwise available to us. Therefore, I have defined abstract classes for LocalRevConnRegistry, RevConnRegistry, etc so that these two entities can be stored in the form of a ThreadLocalRegistry and accessed later. We briefly touched upon this in [this comment](https://github.com/envoyproxy/envoy/issues/33320#issuecomment-2299569235) and on slack. In summary:

1. We have created two thread local entities which get initialized right after the workers are created. They need to be parked somewhere so that they can be accessed later in code paths where only the thread local dispatcher is available. We have parked them with the thread local connection handler and therefore added the abstract classes under Envoy::Network.
2. If the above is not the best approach, what would you suggest?

botengyao replied:
> This thread local registry is then parked with the Connection Handler and is accessed later wherever we need access to the ReverseConnectionManager and ReverseConnectionHandler like here,to initiate reverse conns, here in the reverse conn filter, which accepts reverse conns and in multiple other places where only the thread local dispatcher is otherwise available to us

For filters we can get the slot from  serverFactoryContext from `Server::Configuration::FactoryContext& context` during the filter chain creation, could you try to see if this can avoid the `setConnectionHandler` api changes?

alyssawilk reviewed (commented):
I'm really glad you got this working!  Tossed in a few high level comments but I'm going to assign a first pass reviewer for the rest

botengyao commented on source/common/listener_manager/connection_handler_impl.cc:
```diff
@@ -309,6 +320,37 @@ void ConnectionHandlerImpl::setListenerRejectFraction(UnitFloat reject_fraction)
+                 "Thread local rverse conn registry should not be null.");
+}
+
+void ConnectionHandlerImpl::saveUpstreamConnection(Network::ConnectionSocketPtr&& upstream_socket,
```
We want to store the upstream socket to a dedicated `REVESR_CLUSTER` cluster so that it can be reused when public Envoy wants to establish connections to the on-perm one. And we also wan to save upstream connection here is just for PING and other handshake processes.

The request flow is like:
1. on-prem Envoy -> public Envoy 
2. Then the public Envoy stores the downstream socket to the `REVESR_CLUSTER`, 
3. and also initialize the listener to do RPING 
4. And then the service behind the public Envoy can send request -> normal listener -> filster-chain / router -> the stored socket in REVERSE_CLUSTER.

Am I understanding correctly?

basundhara-c replied:
@botengyao yes, most of the above is correct, the request flow is as follows:

Onprem -> Cloud envoy

1. Onprem initiates reverse connections; it creates http connections to cloud envoy and sends a reverse connection initiation request through it.
2. This request is intercepted by the [reverse conn http filter](https://github.com/envoyproxy/envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/source/extensions/filters/http/reverse_conn/reverse_conn_filter.cc#L85) and cloud envoy stores the sockets with the [ReverseConnectionHandler](https://github.com/envoyproxy/envoy/blob/457e1e985a0ef2c0812f809d34aa3add2d029433/envoy/network/reverse_connection_handler.h#L40). This ReverseConnectionHandler periodically sends RPINGs over all such cached sockets.
3. On cloud envoy, a REVERSE_CONNECTION cluster type is defined and is used for all requests that need to be sent over a reverse connection. When a request arrives and this REVERSE_CONNECTION cluster is picked for the route, we interface with the ReverseConnectionHandler described above, get the cached socket and send the request over it.

botengyao commented on source/common/network/connection_impl.cc:
```diff
@@ -263,6 +278,10 @@ void ConnectionImpl::setDetectedCloseType(DetectedCloseType close_type) {
 }
 
 void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
+  if (connection_reused_ || !ConnectionImpl::ioHandle().isOpen()) {
```
When `connection_reused_` is always true, how do we close the socket?

basundhara-c replied:
@botengyao we don't, in this case, since reverse connections are a very small number of long lived connections. However, we might be able to mark sockets dead in the Listener filter which will be passed these sockets (https://github.com/envoyproxy/envoy/pull/37821), will test it out and update it there.

botengyao replied:
emm, the changes in the connection socket with the customized connection reuse check seem hacky. It could be more clean to create a dedicated close sequence only for the reused connection. Besides, the close can also be initialized by the peer, what will happen in this case? The socket will not close but the remote is closed, will we enter a dead loop when we find we cannot read/write data?

basundhara-c replied:
@botengyao I am trying to find a cleaner way of running the close sequence.

botengyao reviewed (commented):
Thanks @basundhara-c, very interesting feature!
Here is a first pass to kick off the process.

basundhara-c commented:
@botengyao thanks a lot for the suggestion on obtaining the slot from the context! I am trying that out along with an attempt to move the code in [extensions](https://github.com/envoyproxy/envoy/pull/37367) to contrib as much as possible and will be sharing the changes shortly!

basundhara-c commented:
@botengyao we have added a bunch of changes according to your suggestion, namely:
- Removed the getters and setters for connection handler and cluster manager
- The RCManager and RCHandler are now self contained within contrib/ and not parked with the Dispatcher. The filters and other entities obtain them through TLS getters. This has reduced the envoy core code.
- Have added changes required to remove references to extensions code as the reverse connections code is now in contrib/
Will await further suggestions on this PR, thank you in advance!

alyssawilk commented:
@botengyao can you take a look today?
I'm going to remove myself for now just to get it off PR notifier but please add me back once you're done with your pass.

botengyao commented on source/common/json/json_loader.cc:
```diff
@@ -18,5 +18,18 @@ std::vector<uint8_t> Factory::jsonToMsgpack(const std::string& json) {
   return Nlohmann::Factory::jsonToMsgpack(json);
 }
 
+const std::string Factory::listAsJsonString(const std::list<std::string>& items) {
```
These helper methods are used in the extensions, is it possible to add the support incrementally? Also these utility methods need tests.

basundhara-c replied:
I have kept listAsJsonString and removed the others, and have added unit tests for the function

botengyao commented on source/common/listener_manager/connection_handler_impl.cc:
```diff
@@ -338,6 +377,14 @@ ConnectionHandlerImpl::PerAddressActiveListenerDetails::internalListener() {
 
+OptRef<Network::ReverseConnectionListener>
+ConnectionHandlerImpl::PerAddressActiveListenerDetails::reverseConnectionListener() {
+  ENVOY_LOG(debug, "reverseConnectionListener");
```
nit, could avoid this log item, or make it more informative.

basundhara-c replied:
Thanks for pointing this out! Removed the log!

botengyao commented on source/common/listener_manager/listener_impl.cc:
```diff
@@ -591,6 +596,44 @@ ListenerImpl::buildInternalListener(const envoy::config::listener::v3::Listener&
+                    config.name()));
+  }
+  // Reverse connection listener should not bind to port.
+  bind_to_port_ = false;
```
It would be better to add the invalid config check here, it the config contains bind_to_port, we can reject it.

basundhara-c replied:
@botengyao  The bind_to_port_ field is set to true by default in shouldBindToPort(config). The idea here is to set bind_to_port_ to false for any listener with a reverse connection config. Are you suggesting that we make it mandatory for bind_to_port to be explicitly set to false in the listener config for reverse connection listener?

basundhara-c replied:
The reverse_connection_listener_config was indeed missing, my apologies! Added it in listener.proto

botengyao replied:
This is an indicator that the test is missing, could you add a test to cover the config, please?

botengyao reviewed (commented):
Thanks for working on this! Here is another pass, and some changes need at least unit tests to improve the coverage.

/wait

basundhara-c commented on envoy/network/listener.h:
```diff
@@ -134,6 +134,40 @@ class InternalListenerConfig {
+public:
+  virtual ~ReverseConnectionListenerConfig() = default;
+
+  // The source node, cluster and tenant IDs of the local envoy. This is used when the listener is
```
Done

basundhara-c commented on envoy/network/connection_handler.h:
```diff
@@ -303,6 +319,18 @@ class InternalListener : public virtual ConnectionHandler::ActiveListener {
+ */
+class ReverseConnectionListener : public virtual ConnectionHandler::ActiveListener {
+public:
+  virtual void startRCWorkflow(Event::Dispatcher& dispatcher,
```
Done

basundhara-c commented on source/common/listener_manager/connection_handler_impl.h:
```diff
@@ -153,12 +162,13 @@ class ConnectionHandlerImpl : public ConnectionHandler,
   // Declare before its users ActiveListenerDetails.
   std::atomic<uint64_t> num_handler_connections_{};
-  absl::flat_hash_map<uint64_t, std::unique_ptr<ActiveListenerDetails>> listener_map_by_tag_;
+  absl::flat_hash_map<uint64_t, std::shared_ptr<ActiveListenerDetails>> listener_map_by_tag_;
```
Thanks for pointing this out! A unique ptr and a setter as you described works for our used case, and we've made this change!

botengyao reviewed (commented):
Much better! Thanks for the contribution and here is another pass.

/wait

agrawroh commented on source/server/server.cc:
```diff
@@ -765,6 +766,7 @@ absl::Status InstanceBase::initializeOrThrow(Network::Address::InstanceConstShar
   ASSERT(config_.clusterManager());
   xds_manager_ =
       std::make_unique<Config::XdsManagerImpl>(*config_.clusterManager(), validation_context_);
+  listener_manager_->setClusterManagerForWorkers(config_.clusterManager());
```
Looks like there is no `setClusterManagerForWorkers()` anymore.

basundhara-c replied:
Thanks for pointing this out! Fixed some issues caused due to missed code including removal of this and addition of the reverse_connection_listener_config field in listener.proto 
@botengyao

basundhara-c commented on source/common/upstream/upstream_impl.h:
```diff
@@ -273,6 +276,8 @@ class HostDescriptionImplBase : virtual public HostDescription,
   const MonotonicTime creation_time_;
   absl::optional<MonotonicTime> last_hc_pass_time_;
+  // This field is needed to fetch socket from rc_handler for reverse connection.
+  absl::string_view host_id_;
```
Changed this to std::string!

basundhara-c commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2374,6 +2374,17 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
   }
   if (protocols.size() == 1 && protocols[0] == Http::Protocol::Http2 &&
       context_.runtime().snapshot().featureEnabled("upstream.use_http2", 100)) {
+    if (host->cluster().clusterType().has_value() && host->cluster().clusterType()->name() == "envoy.clusters.reverse_connection") {
```
Do you mean that we could define an enum with specific custom cluster types? From my understanding, envoy::config::cluster::v3::Cluster::CustomClusterType is used to define a custom cluster type which can be implemented in an extension without needing to list it among known cluster types in envoy core.

botengyao replied:
yeah, that should work.

botengyao reviewed (commented):
Hi @phlax, do you have insights why the ci doesn't work for this PR?
It is showing `Run # Check if the merge commit SHA is not null
Merge commit information not found for pull request 37368.`

botengyao commented on source/common/http/filter_manager.cc:
```diff
@@ -1000,10 +1000,12 @@ void DownstreamFilterManager::sendLocalReply(
     }
-
+  
+    bool reverse_conn_force_local_reply = 
```
We perfer latching the runtime value to avoid the cost to get it every time sending a local reply.

botengyao commented on source/common/http/filter_manager.cc:
```diff
@@ -1000,10 +1000,12 @@ void DownstreamFilterManager::sendLocalReply(
     // We only prepare a local reply to execute later if we're actively
     // invoking filters to avoid re-entrant in filters.
-    if (state_.filter_call_state_ & FilterCallState::IsDecodingMask) {
+    if (!reverse_conn_force_local_reply && state_.filter_call_state_ & FilterCallState::IsDecodingMask) {
```
Actually a runtime guard has a 6 month deletion policy, and it is usually used to guard the data plane behavior change. I think here you want a parameter to tell if this is going to be a local reply on a reverse connection. Could you point me where you detect this use case, and I assume it is in your HTTP filter extension.

basundhara-c replied:
Yes, this is related to the http filter. Basically, the reverse connection changes are dependent on the guard http_filter_avoid_reentrant_local_reply, which is now deprecated. Without it, when the responder [accepts the reverse connection](https://github.com/envoyproxy/envoy/blob/568ca983adb9bde2fe2dc9d726549e251fb2059d/contrib/reverse_connection/filters/http/source/reverse_conn_filter.cc#L85) and calls sendLocalReply() to ACK the connection, the filter manager prepares a local reply to be sent after the filter execution returns. This results in a segfault as the connection is closed. To allow the reverse conn filter to send a local reply similar to http_filter_avoid_reentrant_local_reply, I added the runtime guard. Would you suggest adding a flag instead?

botengyao reviewed (commented):
Thanks for the dedicated effort! Looks good in high level except the runtime guard and connection close sequence.

/wait

phlax commented:
@botengyao 

> **This branch has conflicts that must be resolved**

jmarantz commented:
/wait (Boteng has unresolved comments)

basundhara-c commented:
@botengyao I wanted to discuss the connection close sequence. The main issue is on the initiator side where [we pass the socket to the listener](https://github.com/nutanix/envoyproxy-envoy/blob/ae22e8a8f5d779146bdadcf5805fdc2d9be03b12/contrib/reverse_connection/bootstrap/source/reverse_connection_initiator.cc#L309) which will henceforth own it and respond to RPINGs on the socket. Therefore, after the [function](https://github.com/nutanix/envoyproxy-envoy/blob/ae22e8a8f5d779146bdadcf5805fdc2d9be03b12/contrib/reverse_connection/bootstrap/source/reverse_connection_initiator.cc#L247) returns, the connection object goes out of scope and the destructor is called, which in turn calls close(). However, the socket has been moved, and socket_ is null. We could remove the following check from close() to make it cleaner:

void ConnectionImpl::close(ConnectionCloseType type) {
  if (connection_reused_ || !ioHandle().isOpen()) {
    ENVOY_CONN_LOG_EVENT(
        debug, "connection_closing",
        "Not closing conn, conn is to be reused or IO handle is closed: connection_reused_:{}",
        *this, connection_reused_);
    // return;
  }
  ...}
  
  However, we would need to check for a null socket here and therefore the close() call would return from this point:

void ConnectionImpl::close(ConnectionCloseType type) {
  if (socket_== nullptr || !socket_->isOpen()) {
    ENVOY_CONN_LOG_EVENT(debug, "connection_closing", "Not closing conn, socket object is null or socket is not open", *this);
    return;
  }
 ...
}

Thereafter, in the listener filter owning the socket, we could handle the socket closure in [the onClose() function](https://github.com/nutanix/envoyproxy-envoy/blob/272eec8782cb0749eb0e7012d045a60af429487d/contrib/reverse_connection/filters/listener/source/reverse_connection.cc#L36).
I look forward to your suggestions on this!

botengyao commented on envoy/network/connection.h:
```diff
@@ -318,6 +318,27 @@ class Connection : public Event::DeferredDeletable,
+   * Set the flag connection_reused_ to value. The flag connection_reused_
+   * indicates whether the client connection is reused.
+   */
+  virtual void setConnectionReused(bool value) PURE;
```
There are 2 connection reused represent 2 use cases, and one is the client case, and one is server connection, will the server connection also set connection reused? would be better to add more comments to explain the usage.

basundhara-c replied:
Added more detailed comments to explain the usage

botengyao replied:
would `setClientConnectionReused()` be better? And also change the underlying `connection_reuse_` to `client_connection_reuse_`?

botengyao reviewed (commented):
Thanks! The socket null check makes sense to me. 

I am actually also thinking if we need to create a new connection class called `ReverseConnection` which overrides some methods, so that you can avoid multiple parameters check, but that could introduce more changes.

/wait

basundhara-c commented:
@botengyao I added the following in the last commit:
- Added a flag "reverse_conn_force_local_reply_" to send a local reply for reverse connections instead of the runtime guard
- Changed the connection close sequence to what I described [here](https://github.com/envoyproxy/envoy/pull/37368#issuecomment-2616926798) 
- Added unit tests for the buildReverseConnectionListener() function

However, due to a change in my VM, I missed out on the DCO sign off, and also wanted to add some nits I missed (a mock  for setReverseConnForceLocalReply and the runtime guard code removal). Is it okay if I rebase to amend the last commit and add the DCO signoff?

botengyao commented:
@basundhara-c, yes, could amend to add DCO to your last commit, and update the branch.

basundhara-c commented:
> @basundhara-c, yes, could amend to add DCO to your last commit, and update the branch.

Done!

botengyao commented on source/common/network/connection_impl.cc:
```diff
@@ -1012,6 +1029,18 @@ ClientConnectionImpl::ClientConnectionImpl(
   }
 }
 
+// Constructor to create "clientConnection" object from an existing downstream connection
```
s/connection/socket

botengyao reviewed (commented):
Almost there and thanks for working on this!

/wait

botengyao commented:
@basundhara-c, also this PR needs a main merge.

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -263,6 +271,10 @@ void ConnectionImpl::setDetectedCloseType(DetectedCloseType close_type) {
 }
 
 void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
+  if (connection_reused_ || !ConnectionImpl::ioHandle().isOpen()) {
```
The distinction was made earlier to distinguish between two connections:
- The connection created by initiator envoy, from which the socket is passed to and owned by the listener. The socket is moved from the connection object. The connection_reused_ was checked at the start of close sequences and the workflow returns 
- The connection accepted by the responder, for which the socket should not be closed. 

However with [this](https://github.com/envoyproxy/envoy/blob/0e26d535ecc5853f4be476666f0e25b2c399a302/source/common/network/connection_impl.cc#L271) check now at the start of the connection close sequence, I have compressed it to one flag which is set for the connection accepted by responder envoy.

adisuissa commented:
CI failures seem to be relevant.
/wait

botengyao commented on source/common/network/connection_impl.h:
```diff
@@ -62,6 +62,13 @@ class ConnectionImpl : public ConnectionImplBase, public TransportSocketCallback
+  // New Method added to retrieve the connection socket.
+  // so that, new connection can be created with existing connection socket.
+  // This is required for reverse connections.
+  const ConnectionSocketPtr& getSocket() const override { return socket_; }
```
I noticed here a const socket ptr reference is returned, and later a `const_cast` is used to get the ptr by force. It would be better to directly return the ptr by value, and then `std::move()`

basundhara-c replied:
I changed ConnectionSocketPtr to a unique_ptr since it was causing issues in the DLB tests and removed the const here

botengyao commented on source/common/network/connection_impl.h:
```diff
@@ -143,6 +150,11 @@ class ConnectionImpl : public ConnectionImplBase, public TransportSocketCallback
+  // Used on the responder envoy to mark an active connection accepted by a listener which will
+  // be used as a reverse connection. The socket for such a connection is closed upon draining
+  // of the owning listener.
+  bool reuse_connection_ = false;
```
Consider moving this to the line bellow of `bool transport_wants_read_ : 1;` and initialize it as `false` in the `ConnectionImpl` construtor.

basundhara-c replied:
Done!

botengyao commented on source/common/listener_manager/active_tcp_listener.cc:
```diff
@@ -55,6 +55,7 @@ ActiveTcpListener::~ActiveTcpListener() {
     ASSERT(active_connections != nullptr);
     auto& connections = active_connections->connections_;
     while (!connections.empty()) {
+      connections.front()->connection_->setConnectionReused(false);
```
Please add a comment here to explain why setting false is needed here. Like by default the normal close will never close the socket if connection is reused.

basundhara-c replied:
Done!

botengyao reviewed (commented):
Thanks for working on this, and left several more comments. Otherwise this is much cleaner than before.

/wait

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -145,7 +143,8 @@ void ConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
 
 void ConnectionImpl::close(ConnectionCloseType type) {
-  if (!socket_->isOpen()) {
+  if (socket_== nullptr || !socket_->isOpen()) {
```
Done

basundhara-c commented:
/retest

basundhara-c commented:
/coverage

RyanTheOptimist commented:
/wait
on CI passing

agrawroh commented:
/retest

srekkas commented:
Hi. This is a highly anticipated feature for us. What’s the status of it?

basundhara-c commented:
/coverage

basundhara-c commented:
/retest

basundhara-c commented:
/retest

basundhara-c commented:
/coverage

botengyao reviewed (commented):
Thanks Basundhara for this epic contribution, excited to see this is happenning in Envoy. And here are more comments and mainly on the `socket_` management and nits.

agrawroh commented:
/retest

agrawroh commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2425,6 +2425,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
       context_.runtime().snapshot().featureEnabled("upstream.use_http2", 100)) {
+    if (host->cluster().clusterType().has_value() &&
+        host->cluster().clusterType()->name() == "envoy.clusters.reverse_connection") {
+      ENVOY_LOG_MISC(debug, "Allocating reverse connection conn pool");
```
@botengyao I was giving it a try but I think the way we wire up the Logger, it doesn't work if we go with `ENVOY_LOG`.

I see these errors:

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
source/common/upstream/cluster_manager_impl.cc:2404:7: error: use of undeclared identifier '__log_do_not_use_read_comment'
 2404 |       ENVOY_LOG(debug, "allocating reverse connection pool");
      |       ^
./source/common/common/logger.h:561:51: note: expanded from macro 'ENVOY_LOG'
  561 | #define ENVOY_LOG(LEVEL, ...) ENVOY_LOG_TO_LOGGER(ENVOY_LOGGER(), LEVEL, ##__VA_ARGS__)
      |                                                   ^
./source/common/common/logger.h:547:24: note: expanded from macro 'ENVOY_LOGGER'
  547 | #define ENVOY_LOGGER() __log_do_not_use_read_comment()
      |                        ^
source/common/upstream/cluster_manager_impl.cc:2404:7: error: use of undeclared identifier '__log_do_not_use_read_comment'
./source/common/common/logger.h:561:51: note: expanded from macro 'ENVOY_LOG'
  561 | #define ENVOY_LOG(LEVEL, ...) ENVOY_LOG_TO_LOGGER(ENVOY_LOGGER(), LEVEL, ##__VA_ARGS__)
      |                                                   ^
./source/common/common/logger.h:547:24: note: expanded from macro 'ENVOY_LOGGER'
  547 | #define ENVOY_LOGGER() __log_do_not_use_read_comment()
      |                        ^
source/common/upstream/cluster_manager_impl.cc:2412:9: error: use of undeclared identifier '__log_do_not_use_read_comment'
 2412 |         ENVOY_LOG(error, "failed to create reverse connection pool. Cannot find a factory "
      |         ^
./source/common/common/logger.h:561:51: note: expanded from macro 'ENVOY_LOG'
  561 | #define ENVOY_LOG(LEVEL, ...) ENVOY_LOG_TO_LOGGER(ENVOY_LOGGER(), LEVEL, ##__VA_ARGS__)
      |                                                   ^
./source/common/common/logger.h:547:24: note: expanded from macro 'ENVOY_LOGGER'
  547 | #define ENVOY_LOGGER() __log_do_not_use_read_comment()
      |                        ^
source/common/upstream/cluster_manager_impl.cc:2412:9: error: use of undeclared identifier '__log_do_not_use_read_comment'
./source/common/common/logger.h:561:51: note: expanded from macro 'ENVOY_LOG'
  561 | #define ENVOY_LOG(LEVEL, ...) ENVOY_LOG_TO_LOGGER(ENVOY_LOGGER(), LEVEL, ##__VA_ARGS__)
      |                                                   ^
./source/common/common/logger.h:547:24: note: expanded from macro 'ENVOY_LOGGER'
  547 | #define ENVOY_LOGGER() __log_do_not_use_read_comment()
      |                        ^
4 errors generated.


source/common/tcp_proxy/tcp_proxy.h
source/common/upstream/cluster_manager_impl.cc
```

botengyao replied:
You would need to inherit the logger id from the factory class, but seems wired for a factory, and can we avoid these kind of logs and cover it by tests?

basundhara-c commented on source/common/http/async_client_impl.h:
```diff
@@ -156,6 +156,11 @@ class AsyncStreamImpl : public virtual AsyncClient::Stream,
   StreamInfo::StreamInfoImpl& streamInfo() override { return stream_info_; }
 
+  void setReverseConnForceLocalReply(bool value) override {
+    ENVOY_LOG_MISC(
```
Using ENVOY_LOG is causing the issue @agrawroh mentioned in the comment

basundhara-c commented on source/extensions/clusters/common/logical_host.cc:
```diff
@@ -92,5 +92,10 @@ Upstream::Host::CreateConnectionData LogicalHost::createConnection(
       std::make_shared<RealHostDescription>(address, shared_from_this()));
 }
 
+void RealHostDescription::setHostId(const std::string host_id) {
```
Done

basundhara-c commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -570,6 +570,10 @@ class Filter : public Network::ReadFilter,
     }
+
+    void setReverseConnForceLocalReply(bool value) override {
+      ENVOY_LOG_MISC(error, "Cannot set value {}.", value);
```
Done

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -118,8 +118,6 @@ ConnectionImpl::ConnectionImpl(Event::Dispatcher& dispatcher, ConnectionSocketPt
 }
 
 ConnectionImpl::~ConnectionImpl() {
-  ASSERT(!socket_->isOpen() && delayed_close_timer_ == nullptr,
```
@agrawroh took care of this in his commit!

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -267,7 +272,8 @@ void ConnectionImpl::setDetectedCloseType(DetectedCloseType close_type) {
 
 void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
-  if (!socket_->isOpen()) {
+
```
@agrawroh took care of this in his commit!

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -304,8 +313,10 @@ void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
 #endif
   }
 
-  // It is safe to call close() since there is an IO handle check.
```
same as above!

agrawroh commented:
/retest

basundhara-c commented:
/coverage

agrawroh commented:
/retest

basundhara-c commented:
/coverage

basundhara-c commented:
/retest

basundhara-c commented:
/retest

basundhara-c commented:
/coverage

basundhara-c commented:
/retest

basundhara-c commented:
/coverage

basundhara-c commented:
/coverage

basundhara-c commented:
/retest

botengyao commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2399,6 +2399,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
       context_.runtime().snapshot().featureEnabled("upstream.use_http2", 100)) {
+    if (host->cluster().clusterType().has_value() &&
+        host->cluster().clusterType()->name() == "envoy.clusters.reverse_connection") {
+      ENVOY_LOG_MISC(debug, "allocating reverse connection pool");
```
If you really want to keep this log, letting ProdClusterManagerFactory inherit `protected Logger::Loggable<Logger::Id::pool>` could work.

botengyao commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2399,6 +2399,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
+            dispatcher, context_.api().randomGenerator(), server_.singletonManager(), host,
+            priority, options, transport_socket_options, state, origin, alternate_protocols_cache);
+      } else
+        ENVOY_LOG_MISC(error,
```
Okay, if it cannot find the factory, should we return here? Since the cluster type is already reverse_connection.

agrawroh replied:
Yeah, I think that would be much cleaner

botengyao commented on source/common/http/http2/conn_pool.h:
```diff
@@ -37,6 +38,26 @@ allocateConnPool(Event::Dispatcher& dispatcher, Random::RandomGenerator& random_
+/**
+ * Abstract class for allocating reverse connection pools.
+ */
+class ReverseConnPoolFactory : public Config::UntypedFactory {
```
Can we use `TypedFactory` and distinguish the extension by the proto message?

agrawroh replied:
Per the current state, it doesn't need any protobuf configuration. The configuration is handled separately through `ReverseConnectionListenerConfig`. Do you think we should add the blank proto and switch to `TypedFactory`?

I don't know the advantages of using one over the other. Will defer it to you :)

botengyao commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2399,6 +2399,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
+        host->cluster().clusterType()->name() == "envoy.clusters.reverse_connection") {
+      ENVOY_LOG_MISC(debug, "allocating reverse connection pool");
+      auto* factory = Config::Utility::getFactoryByName<Http::Http2::ReverseConnPoolFactory>(
+          "envoy.http.reverse_conn.default");
```
It is looking for the factory by name, and this will limit this code path to a specific cluster with this name, could you use the `Factory* getFactory(const ProtoMessage& message)` by type? And then this lookup will be extendable. See above comments for the `ReverseConnPoolFactory`

agrawroh replied:
Okay, let me try to convert it to the `TypedFactory` as suggested above and switch the lookup.

botengyao reviewed (commented):
Thanks for working on this, here is another pass. It is almost there!

/wait

agrawroh commented on envoy/network/filter.h:
```diff
@@ -440,6 +440,12 @@ class ListenerFilter {
+   * Called when the connection is closed. Only the current filter that has stopped filter
+   * chain iteration will get the callback.
+   */
+  virtual void onClose() {};
```
If we make it PURE then it would require all listener filters to implement it. I think this could be a breaking change. What do you think?

agrawroh commented on source/common/listener_manager/active_stream_listener_base.cc:
```diff
@@ -128,7 +128,9 @@ void ActiveTcpConnection::onEvent(Network::ConnectionEvent event) {
       event == Network::ConnectionEvent::RemoteClose) {
     stream_info_->setDownstreamTransportFailureReason(connection_->transportFailureReason());
-    active_connections_.listener_.removeConnection(*this);
+    if (!connection_->isConnectionReused()) {
```
IIRC the overall sequence is like this:
1. Connection is marked as reused.
2. `moveSocket()` is called to transfer the ownership of the socket.
3. Connection remains in the active list until the transfer is complete.
4. Connection is deferred deleted after the transfer.
5. Filters are cleaned up during deferred deletion.

cc @basundhara-c Is this understanding correct?

botengyao replied:
The defer-deletion happens in the removeConnection（）， if this is gated, then what is the purpose to raise close event in your 'moveScocket'？

agrawroh replied:
Yeah, I think the right thing to do here would be to remove that gating and let the connection cleanup happen normally.

basundhara-c commented:
/retest

botengyao commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2399,6 +2399,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
+        host->cluster().clusterType()->name() == "envoy.clusters.reverse_connection") {
+      ENVOY_LOG(debug, "allocating reverse connection pool");
+      auto* factory = Config::Utility::getFactory<Http::Http2::ReverseConnPoolFactory>(
+          envoy::extensions::upstreams::http::reverse_conn::v3::ReverseConnPoolProto::
```
Rather than a default instance, you would need to pass the `ClusterType()` to `getFactory` or `ClusterType()->typed_config()` to `getFactoryByType()`, and then define a customized connection pool factory in this PR: https://github.com/envoyproxy/envoy/pull/37931. So that it is extendable and it is using the typed_config to search the factory, and no default proto is needed.

botengyao commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -2399,6 +2399,20 @@ Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
+          envoy::extensions::upstreams::http::reverse_conn::v3::ReverseConnPoolProto::
+              default_instance());
+      if (factory == nullptr) {
+        ENVOY_LOG_MISC(error, "factory not found for reverse connection pool");
```
nit: ENVOY_LOG

botengyao reviewed (commented):
Thanks! To format you can run `bazel run //tools/code_format:check_format -- fix`

/wait

agrawroh commented:
/retest

botengyao commented on source/common/network/connection_impl.cc:
```diff
@@ -285,6 +292,36 @@ void ConnectionImpl::setDetectedCloseType(DetectedCloseType close_type) {
   detected_close_type_ = close_type;
 }
 
+ConnectionSocketPtr ConnectionImpl::moveSocket() {
```
I also noticed there's no test coverage for this part. Please add a unit test in connection_impl_test. Additionally, it should be possible to simulate the move socket behavior in an integration test. You could create a simple test network filter and invoke moveSocket to verify that the behavior matches expectations.

Given that this involves data plane changes and potential lifetime issues, those are the primary concerns here.

basundhara-c replied:
Added a simple unit test for this behavior in ConnectionImplTest, do let me know if it suffices!

botengyao reviewed (commented):
/wait

agrawroh commented:
/retest

basundhara-c commented on source/common/network/connection_impl.h:
```diff
@@ -260,6 +268,11 @@ class ConnectionImpl : public ConnectionImplBase, public TransportSocketCallback
+  // Used on the responder envoy to mark an active connection accepted by a listener which will
+  // be used as a reverse connection. The socket for such a connection is closed upon draining
+  // of the owning listener.
+  bool reuse_connection_ : 1;
```
Changed this to reuse_socket_

basundhara-c commented on source/common/network/connection_impl.cc:
```diff
@@ -188,7 +190,12 @@ void ConnectionImpl::closeInternal(ConnectionCloseType type) {
       // We aren't going to wait to flush, but try to write as much as we can if there is pending
       // data.
-      transport_socket_->doWrite(*write_buffer_, true);
+      if (reuse_connection_) {
```
Added!

yanavlasov commented:
I have posted a question to the Issue #33320. I think you can likely significantly simplify this design.

botengyao commented:
waiting on addressing Yan's feedback.

/wait
